### PR TITLE
ztp: Cleanup multinode siteconfig schema

### PR DESCRIPTION
**Pull request by lack**
- ztp: Compute SiteConfig NumMasters and NumWorkers from node roles
- ztp: deprecate SiteConfig clusters.clusterType

### CODE REVIEW

File: ztp/ran-crd/site-config-crd.yaml [hunk 1]
```diff
@@ -78,21 +78,16 @@ spec:
                         description: BaseDomain is the base domain to which the cluster should belong.
                         type: string
                       clusterType:
-                        description: clusterType either standard or sno
+                        deprecated: true
+                        description: clusterType either standard or sno (Ignored: Computed based on number of master nodes defined)
```
Comment (lack) [inline ztp/ran-crd/site-config-crd.yaml:82]: Open question to @serngawy and @imiller0 : Should I keep these fields in as "deprecated", or should I just remove them from the CRD altogether?

File: ztp/siteconfig-generator/siteConfig/siteConfig_test.go [hunk 2]
```diff
@@ -88,21 +75,79 @@ spec:
...
-	// Validate ClusterType
+	// Validate Disk Encryption type
 	assert.Equal(t, siteConfig.Spec.Clusters[0].DiskEncryption.Type, "nbde")
 	assert.Equal(t, siteConfig.Spec.Clusters[1].DiskEncryption.Type, "none")
 	assert.Equal(t, siteConfig.Spec.Clusters[2].DiskEncryption.Type, "none")
+}
+
+func TestRoleCounters(t *testing.T) {
+	input := `
+apiVersion: ran.openshift.io/v1
```
Comment (serngawy) [inline ztp/siteconfig-generator/siteConfig/siteConfig_test.go:97]: There is a test case where the node role not set, it should be master by default

File: ztp/siteconfig-generator/siteConfig/siteConfig.go [hunk 3]
```diff
@@ -91,34 +91,49 @@ type Clusters struct {
...
+	rv.NumWorkers = 0
+	for _, node := range rv.Nodes {
+		if node.Role == "master" {
+			rv.NumMasters += 1
+		} else {
+			rv.NumWorkers += 1
+		}
+	}
+	// Autodetect ClusterType based on the node counts
+	if rv.NumMasters == 1 && rv.NumWorkers == 0 {
```
Comment (serngawy) [inline ztp/siteconfig-generator/siteConfig/siteConfig.go:132]: I believe we should raise error if the num of nodes not match requirements similar to AI. 
- standard cluster num of masters must be 3 
- SNO cluster num of master must be 1

File: ztp/siteconfig-generator/siteConfig/siteConfig_test.go [hunk 4]
```diff
@@ -88,21 +75,79 @@ spec:
...
+kind: SiteConfig
+spec:
+  clusters:
+  - clusterName: "ignore-user-supplied-values"
+    numMasters: 5
```
Comment (serngawy) [inline ztp/siteconfig-generator/siteConfig/siteConfig_test.go:102]: if the user set num of masters and workers that mean he know what he is doing, we should check the nodes he set and raise error if the num of masters/workers didn't match.

**Review by serngawy — CHANGES_REQUESTED** [review]

File: ztp/siteconfig-generator/siteConfig/siteConfig_test.go (hunk 4 above)
Comment (lack) [reply ztp/siteconfig-generator/siteConfig/siteConfig_test.go:102]: No, the user should never set numMasters or numWorkers.  The point of this change is that there is no reason to do so and we should remove that field from the CRD entirely.

File: ztp/siteconfig-generator/siteConfig/siteConfig.go [hunk 5]
```diff
@@ -91,34 +91,49 @@ type Clusters struct {
...
 	}
 
 	out := defaults
 	err := unmarshal(&out)
 	*rv = Clusters(out)
+	// Tally master and worker counts based on node roles
+	rv.NumMasters = 0
+	rv.NumWorkers = 0
+	for _, node := range rv.Nodes {
+		if node.Role == "master" {
```
[the remark this answers is not in the document]
Comment (lack) [reply ztp/siteconfig-generator/siteConfig/siteConfig.go:125]: This assumes the nodes have already been initialized properly, but I take your point.

[the remark this answers is not in the document]
Comment (lack) [reply ztp/siteconfig-generator/siteConfig/siteConfig.go:125]: Hmmm... In order to be future-proof (where there may be more non-master roles like "worker-du" or variants like that), I think I have a solution that works - If Role is "master" or "", assume master.  Any other non-zero string is a worker of some kind.

File: ztp/ran-crd/site-config-crd.yaml (hunk 1 above)
Comment (lack) [reply ztp/ran-crd/site-config-crd.yaml]: Removed now.

File: ztp/ran-crd/site-config-crd.yaml [hunk 6]
```diff
@@ -77,23 +77,6 @@ spec:
...
-                        description: Number of masters for the control plane.
-                        type: integer
-                        default: 1
-                        enum:
-                          - 1
-                          - 3
-                      numWorkers:
-                        description: Number of workers for the control plane.
-                        type: integer
                       clusterProfile:
```
Comment (serngawy) [inline ztp/ran-crd/site-config-crd.yaml:80]: would u clean the cluster Profile as well, no use for it.

**Review by serngawy — COMMENTED** [review]
sounds good, if u can remove the clusterProfile as well. if not we can do it in another PR

File: ztp/ran-crd/site-config-crd.yaml (hunk 6 above)
Comment (lack) [reply ztp/ran-crd/site-config-crd.yaml:80]: Sure!

**Review by serngawy — APPROVED** [review]

**Pull request merged**